# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1683590317 -7200
# Parent  b0d352ff3bea6a318fa57037606d78cb818f2424
Feature: Re-implementation of the two-line view with columns for BB 115.

diff --git a/mail/app/profile/all-thunderbird.js b/mail/app/profile/all-thunderbird.js
--- a/mail/app/profile/all-thunderbird.js
+++ b/mail/app/profile/all-thunderbird.js
@@ -328,16 +328,17 @@ pref("mail.ui-rdf.version", 0);
 pref("mail.showCondensedAddresses", true); // show the friendly display name for people I know
 // 1: Show display name, 2: Show e-mail address, 3: Show both.
 pref("mail.addressDisplayAuthor", 1);
 pref("mail.addressDisplayRecipients", 1);
 
 pref("mailnews.attachments.display.start_expanded", false);
 // hidden pref for changing how we present attachments in the message pane
 pref("mail.pane_config.dynamic",            0);
+pref("mail.pane_config.multiline_all", false);
 pref("mailnews.reuse_thread_window2",     true);
 pref("editor.singleLine.pasteNewlines", 4);  // substitute commas for new lines in single line text boxes
 pref("editor.CR_creates_new_p", true);
 pref("mail.compose.default_to_paragraph", true);
 
 // If true, when pasting a URL, paste the Open Graph / Twitter Card details
 // we can extract from the URL instead.
 pref("mail.compose.add_link_preview", false);
diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -2933,16 +2933,17 @@ var threadPane = {
 
   columns: getDefaultColumns(gFolder),
 
   async init() {
     quickFilterBar.init();
 
     this.setUpTagStyles();
     Services.prefs.addObserver("mailnews.tags.", this);
+    Services.prefs.addObserver("mail.pane_config.multiline_all", this);
 
     Services.obs.addObserver(this, "addrbook-displayname-changed");
 
     // Ensure TreeView and its classes are properly defined.
     await customElements.whenDefined("tree-view-table-row");
 
     threadTree = document.getElementById("threadTree");
     this.treeTable = threadTree.table;
@@ -2962,16 +2963,20 @@ var threadPane = {
       this,
       "selectDelay",
       "mailnews.threadpane_select_delay",
       null,
       (name, oldValue, newValue) => (threadTree.dataset.selectDelay = newValue)
     );
     threadTree.dataset.selectDelay = this.selectDelay;
 
+    // Setup multi-line.
+    let multiline = Services.prefs.getBoolPref("mail.pane_config.multiline_all", false);
+    customElements.get("thread-row").MULTILINE = multiline;
+
     window.addEventListener("uidensitychange", () => {
       this.densityChange();
       threadTree.invalidate();
     });
     this.densityChange();
 
     XPCOMUtils.defineLazyGetter(this, "notificationBox", () => {
       let container = document.getElementById("threadPaneNotificationBox");
@@ -3078,17 +3083,24 @@ var threadPane = {
           threadTree.dispatchEvent(new CustomEvent("select"));
         }
         break;
     }
   },
 
   observe(subject, topic, data) {
     if (topic == "nsPref:changed") {
-      this.setUpTagStyles();
+      if (data == "mail.pane_config.multiline_all") {
+        let multiline = Services.prefs.getBoolPref("mail.pane_config.multiline_all", false);
+        customElements.get("thread-row").MULTILINE = multiline;
+        this.densityChange();
+      } else {
+        this.setUpTagStyles();
+      }
+      threadTree.invalidate();
     } else if (topic == "addrbook-displayname-changed") {
       threadTree.invalidate();
     }
   },
 
   _onDoubleClick(event) {
     if (event.target.closest("button")) {
       // Prevent item activation if double click happens on a button inside the
@@ -3438,25 +3450,25 @@ var threadPane = {
    */
   densityChange() {
     // The class ThreadRow can't be referenced because it's declared in a
     // different scope. But we can get it from customElements.
     let rowClass = customElements.get("thread-row");
     let cardClass = customElements.get("thread-card");
     switch (UIDensity.prefValue) {
       case UIDensity.MODE_COMPACT:
-        rowClass.ROW_HEIGHT = 18;
+        rowClass.ROW_HEIGHT = rowClass.MULTILINE ? 36 : 18;
         cardClass.ROW_HEIGHT = 40;
         break;
       case UIDensity.MODE_TOUCH:
-        rowClass.ROW_HEIGHT = 32;
+        rowClass.ROW_HEIGHT = rowClass.MULTILINE ? 64 : 32;
         cardClass.ROW_HEIGHT = 52;
         break;
       default:
-        rowClass.ROW_HEIGHT = 26;
+        rowClass.ROW_HEIGHT = rowClass.MULTILINE ? 52 : 26;
         cardClass.ROW_HEIGHT = 46;
         break;
     }
   },
 
   /**
    * Store the current thread tree selection.
    */
@@ -4445,27 +4457,37 @@ var folderListener = {
   },
 };
 
 /**
  * Custom element for rows in the thread tree.
  */
 customElements.whenDefined("tree-view-table-row").then(() => {
   class ThreadRow extends customElements.get("tree-view-table-row") {
-    static ROW_HEIGHT = 22;
+    static ROW_HEIGHT = 0; // Will be set via density.
+    static MULTILINE = false;
 
     connectedCallback() {
       if (this.hasConnected) {
         return;
       }
 
       super.connectedCallback();
 
       this.setAttribute("draggable", "true");
       this.appendChild(threadPane.rowTemplate.content.cloneNode(true));
+
+      let table = this.parentNode.parentNode;
+      if (ThreadRow.MULTILINE) {
+        this.setAttribute("multiline", "true");
+        table.setAttribute("multiline", "true");
+      } else {
+        this.removeAttribute("multiline");
+        table.removeAttribute("multiline");
+      }
     }
 
     get index() {
       return super.index;
     }
 
     set index(index) {
       super.index = index;
diff --git a/mail/base/content/mailWindowOverlay.js b/mail/base/content/mailWindowOverlay.js
--- a/mail/base/content/mailWindowOverlay.js
+++ b/mail/base/content/mailWindowOverlay.js
@@ -417,16 +417,25 @@ function InitViewLayoutStyleMenu(event, 
       "paneLayout",
       "table"
     ) === "true"
   ) {
     parent
       .querySelector(`[name="tablelayout"]`)
       .setAttribute("checked", "true");
   }
+
+  let multiline = document.getElementById(
+    appmenu ? "appmenu_multilineAllFolders" : "multilineAllFolders"
+  );
+  if (Services.prefs.getBoolPref("mail.pane_config.multiline_all")) {
+    multiline.setAttribute("checked", "true");
+  } else {
+    multiline.removeAttribute("checked");
+  }
 }
 
 /**
  * Called when showing the menu_viewSortPopup menupopup, so it should always
  * be up-to-date.
  */
 function InitViewSortByMenu() {
   let tab = document.getElementById("tabmail")?.currentTabInfo;
@@ -1441,16 +1450,31 @@ function MsgFilters(emailAddress, folder
     }
   } else {
     // Just launch filterList dialog.
     args = { refresh: false, folder };
     MsgFilterList(args);
   }
 }
 
+function ToogleMultilineAll(appmenu) {
+  let multiline = document.getElementById(
+    appmenu ? "appmenu_multilineAllFolders" : "multilineAllFolders"
+  );
+  let isMultiline = Services.prefs.getBoolPref(
+    "mail.pane_config.multiline_all"
+  );
+  if (isMultiline) {
+    multiline.removeAttribute("checked");
+  } else {
+    multiline.setAttribute("checked", "true");
+  }
+  Services.prefs.setBoolPref("mail.pane_config.multiline_all", !isMultiline);
+}
+
 function MsgViewAllHeaders() {
   Services.prefs.setIntPref(
     "mail.show_headers",
     Ci.nsMimeHeaderDisplayTypes.AllHeaders
   );
 }
 
 function MsgViewNormalHeaders() {
diff --git a/mail/base/content/messenger-menubar.inc.xhtml b/mail/base/content/messenger-menubar.inc.xhtml
--- a/mail/base/content/messenger-menubar.inc.xhtml
+++ b/mail/base/content/messenger-menubar.inc.xhtml
@@ -369,16 +369,20 @@
         <menuitem id="messagePaneVertical" type="radio" label="&messagePaneVertical.label;" name="viewlayoutgroup"
                   accesskey="&messagePaneVertical.accesskey;" command="cmd_viewVerticalMailLayout"/>
         <menuseparator id="viewMenuAfterPaneVerticalSeparator"/>
         <menuitem id="messagePaneVertical"
                   type="checkbox"
                   name="tablelayout"
                   data-l10n-id="menu-view-folders-force-table-layout"
                   command="cmd_toggleForceTableLayout"/>
+        <menuitem id="multilineAllFolders"
+                  type="checkbox"
+                  data-l10n-id="menu-multi-line-all-folders"
+                  oncommand="ToogleMultilineAll(false);"/>
         <menuseparator id="viewMenuBeforeShowFolderPaneSeparator"/>
         <menuitem id="menu_showFolderPane" type="checkbox" label="&showFolderPaneCmd.label;"
                   accesskey="&showFolderPaneCmd.accesskey;" command="cmd_toggleFolderPane"/>
         <menuitem id="menu_showFolderPaneCols" type="checkbox" label="&showFolderPaneColsCmd.label;"
                   accesskey="&showFolderPaneColsCmd.accesskey;" command="cmd_toggleFolderPaneCols"/>
         <menuitem id="menu_showMessage" type="checkbox" label="&showMessageCmd.label;" key="key_toggleMessagePane"
                   accesskey="&showMessageCmd.accesskey;" command="cmd_toggleMessagePane"/>
       </menupopup>
diff --git a/mail/components/customizableui/content/panelUI.inc.xhtml b/mail/components/customizableui/content/panelUI.inc.xhtml
--- a/mail/components/customizableui/content/panelUI.inc.xhtml
+++ b/mail/components/customizableui/content/panelUI.inc.xhtml
@@ -356,16 +356,21 @@
                        command="cmd_viewVerticalMailLayout"/>
         <toolbarseparator id="appmenu_viewMenuAfterPaneVerticalSeparator"/>
         <toolbarbutton id="appmenu_forceTableLayout"
                        class="subviewbutton subviewbutton-iconic"
                        type="checkbox"
                        name="tablelayout"
                        data-l10n-id="appmenuitem-force-table-layout"
                        command="cmd_toggleForceTableLayout"/>
+        <toolbarbutton id="appmenu_multilineAllFolders"
+                       class="subviewbutton subviewbutton-iconic"
+                       type="checkbox"
+                       data-l10n-id="menu-multi-line-all-folders"
+                       oncommand="ToogleMultilineAll(true);"/>
         <toolbarseparator id="appmenu_viewMenuBeforeShowFolderPaneSeparator"/>
         <toolbarbutton id="appmenu_showFolderPane"
                        class="subviewbutton subviewbutton-iconic"
                        type="checkbox"
                        label="&showFolderPaneCmd.label;"
                        command="cmd_toggleFolderPane"/>
         <toolbarbutton id="appmenu_showFolderPaneCols"
                        class="subviewbutton subviewbutton-iconic"
diff --git a/mail/locales/en-US/messenger/menubar.ftl b/mail/locales/en-US/messenger/menubar.ftl
--- a/mail/locales/en-US/messenger/menubar.ftl
+++ b/mail/locales/en-US/messenger/menubar.ftl
@@ -108,16 +108,19 @@ menu-view-repair-text-encoding =
 menu-view-folders-toggle-header =
     .label = Folder Pane Header
     .accesskey = F
 
 menu-view-folders-force-table-layout =
     .label = Message List in Table Layout
     .accesskey = T
 
+menu-multi-line-all-folders =
+    .label = Multi-line View On All Folders
+
 ## View / Layout
 
 menu-font-size-label =
     .label = Font Size
     .accesskey = o
 
 menuitem-font-size-enlarge =
     .label = Increase Font Size
diff --git a/mail/themes/shared/mail/threadPane.css b/mail/themes/shared/mail/threadPane.css
--- a/mail/themes/shared/mail/threadPane.css
+++ b/mail/themes/shared/mail/threadPane.css
@@ -309,26 +309,69 @@ tr[is="thread-row"] td > .thread-contain
 tr[is="thread-row"] .subject-line {
   margin-inline-start: calc(16px * var(--thread-level));
   pointer-events: none;
   /* Line height px exception to avoid vertical cut off of characters. This
      should follow and match the density variation height of the row. */
   line-height: 22px;
 }
 
+/* Override rules from tree-listbox.css */
+tr[is="thread-row"][multiline="true"] td div.thread-container,
+tr[is="thread-row"][multiline="true"] td div.subject-line {
+  overflow: visible !important;
+}
+tr[is="thread-row"][multiline="true"] td.subjectcol-column {
+  overflow: visible !important;
+}
+
+/* Move everything in the row up and the subject down */
+:root[uidensity="compact"] tr[is="thread-row"][multiline="true"] td {
+  top: -9px;
+}
+:root[uidensity="compact"] tr[is="thread-row"][multiline="true"] td.subjectcol-column {
+  top: 7px;
+}
+tr[is="thread-row"][multiline="true"] td {
+  top: -12px;
+}
+tr[is="thread-row"][multiline="true"] td.subjectcol-column {
+  top: 10px;
+}
+:root[uidensity="touch"] tr[is="thread-row"][multiline="true"] td {
+  top: -14px;
+}
+:root[uidensity="touch"] tr[is="thread-row"][multiline="true"] td.subjectcol-column {
+  top: 12px;
+}
+
+/* Make the subject column header fixed width and remove the splitter */
+table[is="tree-view-table"][multiline="true"] #subjectCol {
+  width: 12px !important;
+  min-width: 12px !important;
+  border: 1px solid red;
+}
+table[is="tree-view-table"][multiline="true"] #subjectColSplitter {
+  display: none !important;
+}
+table[is="tree-view-table"][multiline="true"] #subjectColButton {
+  content-visibility: hidden;
+  background-position-x: left 0;
+}
+
 tr[is="thread-row"]:not(.children) .subject-line {
   padding-inline-start: 22px;
 }
 
 tr[is="thread-row"] .subject-line img {
   visibility: hidden;
   width: 16px;
   height: 16px;
   vertical-align: sub;
-  margin-inline-end: 3px;
+  margin-inline-end: 1px;
 }
 
 tr[is="thread-card"] .state {
   display: none;
 }
 
 /* Icons variations for message state in subject column */
 
